pkg/controller/jobs/statefulset/statefulset_reconciler.go

```diff
@@ -48,54 +57,62 @@ type Reconciler struct {
 }
 
 func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
-	sts := &appsv1.StatefulSet{}
-	err := r.client.Get(ctx, req.NamespacedName, sts)
-	if err != nil {
-		// we'll ignore not-found errors, since there is nothing to do.
-		return ctrl.Result{}, client.IgnoreNotFound(err)
-	}
-
-	log := ctrl.LoggerFrom(ctx).WithValues("statefulset", klog.KObj(sts))
+	log := ctrl.LoggerFrom(ctx).WithValues("statefulset", klog.KRef(req.Namespace, req.Name))
 	ctx = ctrl.LoggerInto(ctx, log)
 	log.V(2).Info("Reconciling StatefulSet")
 
-	err = r.fetchAndFinalizePods(ctx, sts)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
-
-	return ctrl.Result{}, nil
+	err := r.fetchAndFinalizePods(ctx, req)
+	return ctrl.Result{}, err
 }
 
-func (r *Reconciler) fetchAndFinalizePods(ctx context.Context, sts *appsv1.StatefulSet) error {
+func (r *Reconciler) fetchAndFinalizePods(ctx context.Context, req reconcile.Request) error {
 	podList := &corev1.PodList{}
-	if err := r.client.List(ctx, podList, client.InNamespace(sts.Namespace), client.MatchingLabels{
-		podcontroller.GroupNameLabel: GetWorkloadName(sts.Name),
+	if err := r.client.List(ctx, podList, client.InNamespace(req.Namespace), client.MatchingLabels{
+		podcontroller.GroupNameLabel: GetWorkloadName(req.Name),
 	}); err != nil {
 		return err
 	}
+	if len(podList.Items) == 0 {
+		return nil
+	}
+
+	sts := &appsv1.StatefulSet{}
+	err := r.client.Get(ctx, req.NamespacedName, sts)
+	if client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	if err != nil {
+		sts = nil
+	}
+
 	return r.finalizePods(ctx, sts, podList.Items)
 }
 
 func (r *Reconciler) finalizePods(ctx context.Context, sts *appsv1.StatefulSet, pods []corev1.Pod) error {
+	errs := make([]error, 0, len(pods))
+	for _, pod := range pods {
+		errs = append(errs, r.finalizePod(ctx, sts, &pod))
+	}
```

Comment:
I think we can run in parallel as in other places removing finalizers, like [here](https://github.com/kubernetes-sigs/kueue/blob/3755065cfde22ef480b882e7558aa8cd1bee58de/pkg/controller/jobs/pod/pod_controller.go#L529-L534)